File: android/VIRGINSApp/app/src/main/java/app/virgins/service/SocketService.kt
package app.virgins.service

import io.socket.client.IO
import io.socket.client.Socket
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import org.json.JSONObject
import javax.inject.Inject
import javax.inject.Singleton

sealed class SocketEvent {
    data class NewMessage(val json: JSONObject) : SocketEvent()
    data class TypingStart(val userId: String) : SocketEvent()
    data class TypingStop(val userId: String) : SocketEvent()
    data class NewMatch(val json: JSONObject) : SocketEvent()
    data class MessageRead(val messageId: String) : SocketEvent()
}

@Singleton
class SocketService @Inject constructor() {

    private var socket: Socket? = null

    private val _events = MutableSharedFlow<SocketEvent>(extraBufferCapacity = 64)
    val events: SharedFlow<SocketEvent> = _events.asSharedFlow()

    fun connect(serverUrl: String, token: String) {
        if (socket?.connected() == true) return
        val options = IO.Options.builder()
            .setAuth(mapOf("token" to token))
            .build()
        socket = IO.socket(serverUrl, options).apply {
            on(Socket.EVENT_CONNECT) { }
            on("receive_message") { args ->
                (args.firstOrNull() as? JSONObject)?.let {
                    _events.tryEmit(SocketEvent.NewMessage(it))
                }
            }
            on("typing_start") { args ->
                (args.firstOrNull() as? JSONObject)?.optString("userId")?.let {
                    _events.tryEmit(SocketEvent.TypingStart(it))
                }
            }
            on("typing_stop") { args ->
                (args.firstOrNull() as? JSONObject)?.optString("userId")?.let {
                    _events.tryEmit(SocketEvent.TypingStop(it))
                }
            }
            on("new_match") { args ->
                (args.firstOrNull() as? JSONObject)?.let {
                    _events.tryEmit(SocketEvent.NewMatch(it))
                }
            }
            on("message_read") { args ->
                (args.firstOrNull() as? JSONObject)?.optString("messageId")?.let {
                    _events.tryEmit(SocketEvent.MessageRead(it))
                }
            }
            connect()
        }
    }

    fun disconnect() {
        socket?.disconnect()
        socket = null
    }

    fun joinRoom(conversationId: String) {
        socket?.emit("join_room", conversationId)
    }

    fun leaveRoom(conversationId: String) {
        socket?.emit("leave_room", conversationId)
    }

    fun sendTypingStart(conversationId: String) {
        socket?.emit("typing_start", JSONObject().put("conversationId", conversationId))
    }

    fun sendTypingStop(conversationId: String) {
        socket?.emit("typing_stop", JSONObject().put("conversationId", conversationId))
    }

    fun markRead(messageId: String) {
        socket?.emit("message_read", JSONObject().put("messageId", messageId))
    }

    fun isConnected() = socket?.connected() == true
}

File: ios/VIRGINSApp/Sources/Services/SocketService.swift
import Foundation
import Combine

// NOTE: Requires SocketIO Swift package:
// https://github.com/socketio/socket.io-client-swift
// Add via SPM: .package(url: "https://github.com/socketio/socket.io-client-swift", from: "16.0.0")

import SocketIO

final class SocketService: ObservableObject {
    static let shared = SocketService()

    private var manager: SocketManager?
    private var socket: SocketIOClient?

    // Publishers
    let newMessagePublisher    = PassthroughSubject<ChatMessage, Never>()
    let typingPublisher        = PassthroughSubject<(conversationId: String, userId: String, isTyping: Bool), Never>()
    let newMatchPublisher      = PassthroughSubject<Match, Never>()
    let messageReadPublisher   = PassthroughSubject<(conversationId: String, messageId: String), Never>()

    @Published var isConnected = false

    func connect(token: String) {
        let serverURL = URL(string: "http://localhost:5000")!
        manager = SocketManager(socketURL: serverURL, config: [
            .log(false),
            .compress,
            .connectParams(["token": token]),
            .reconnects(true),
            .reconnectAttempts(5)
        ])
        socket = manager?.defaultSocket

        socket?.on(clientEvent: .connect)    { [weak self] _, _ in self?.isConnected = true }
        socket?.on(clientEvent: .disconnect) { [weak self] _, _ in self?.isConnected = false }

        socket?.on("receive_message") { [weak self] data, _ in
            guard let json = data.first as? [String: Any],
                  let msg = try? JSONDecoder.isoDecoder.decode(ChatMessage.self, from: JSONSerialization.data(withJSONObject: json)) else { return }
            self?.newMessagePublisher.send(msg)
        }

        socket?.on("typing_start") { [weak self] data, _ in
            guard let d = data.first as? [String: String],
                  let cid = d["conversationId"], let uid = d["userId"] else { return }
            self?.typingPublisher.send((cid, uid, true))
        }

        socket?.on("typing_stop") { [weak self] data, _ in
            guard let d = data.first as? [String: String],
                  let cid = d["conversationId"], let uid = d["userId"] else { return }
            self?.typingPublisher.send((cid, uid, false))
        }

        socket?.on("new_match") { [weak self] data, _ in
            guard let json = data.first as? [String: Any],
                  let match = try? JSONDecoder.isoDecoder.decode(Match.self, from: JSONSerialization.data(withJSONObject: json)) else { return }
            self?.newMatchPublisher.send(match)
        }

        socket?.on("message_read") { [weak self] data, _ in
            guard let d = data.first as? [String: String],
                  let cid = d["conversationId"], let mid = d["messageId"] else { return }
            self?.messageReadPublisher.send((cid, mid))
        }

        socket?.connect()
    }

    func disconnect() {
        socket?.disconnect()
        manager = nil
        socket = nil
        isConnected = false
    }

    func joinRoom(_ conversationId: String) {
        socket?.emit("join_room", ["conversationId": conversationId])
    }

    func leaveRoom(_ conversationId: String) {
        socket?.emit("leave_room", ["conversationId": conversationId])
    }

    func sendTypingStart(_ conversationId: String) {
        socket?.emit("typing_start", ["conversationId": conversationId])
    }

    func sendTypingStop(_ conversationId: String) {
        socket?.emit("typing_stop", ["conversationId": conversationId])
    }

    func markRead(conversationId: String, messageId: String) {
        socket?.emit("message_read", ["conversationId": conversationId, "messageId": messageId])
    }
}

private extension JSONDecoder {
    static var isoDecoder: JSONDecoder {
        let d = JSONDecoder()
        d.keyDecodingStrategy = .convertFromSnakeCase
        d.dateDecodingStrategy = .iso8601
        return d
    }
}
